nestkernel/conn_parameter.h

```diff
@@ -145,50 +148,103 @@ class ScalarIntegerParameter : public ConnParameter
 
 
 /**
- * Array parameter, returning values in order.
+ * Array parameter classes, returning double values in order.
  *
  * - The array of values must not be empty
  *   (so return 0 for number_of_values can signal non-array parameter)
  * - Throws exception if more values requested than available.
- *
+ * - The class contains nthread number of pointers (one for each thread)
+ *   to an iterator, which runs over the parameters initialised in an array.
+ *   Each pointer is moved along the paramtere array by the function
+ *   value_double(), which returns the current parameter value and moves the
+ *   pointer to the subsequent position.
+ * - All parameters are  doubles, thus calling the function value_int()
+ *   throws an error.
  */
-class ArrayParameter : public ConnParameter
+
+class ArrayDoubleParameter : public ConnParameter
 {
 public:
-  ArrayParameter( const std::vector< double >& values )
-    : values_( values )
-    , next_( values_.begin() )
+  ArrayDoubleParameter( const std::vector< double >& values, const size_t nthreads )
+    : values_( &values )
+    , next_( nthreads, values_->begin() )
   {
   }
 
   size_t
   number_of_values() const
   {
-    return values_.size();
+    return values_->size();
   }
 
-  // double value(index sgid, index tgid, librandom::RngPtr&) const
   double
-  value_double( index, index, librandom::RngPtr& ) const
-
+  value_double( thread tid, librandom::RngPtr& ) const
   {
-    // return values_[sgid];
-    if ( next_ != values_.end() )
-      return *next_++;
+    if ( next_[ tid ] != values_->end() )
+      return *next_[ tid ]++;
     else
       throw KernelException( "Parameter values exhausted." );
   }
   long_t
-  value_int( index, index, librandom::RngPtr& ) const
+  value_int( thread, librandom::RngPtr& ) const
   {
     throw KernelException( "ConnParameter calls value function with false return type." );
   }
 
 private:
-  std::vector< double > values_;
-  mutable std::vector< double >::iterator next_;
+  const std::vector< double >* values_;
```

Comment:
I am concerned about the pointer to vector here, fearing a memory leak or worse. In the pertaining `create()` method we dereference the datum twice (`**dvd`) to access the data in the vector. Is there any risk that the actual data would be destroyed while we here still hold a pointer to it? Or that we mess up the reference count of the datum, so it would not be freed? Since these arrays may be large, we might see noticeable memory leaks. We need to look at this carefully. The same applies to the `ArrayIntegerParameter`.